src/main/kotlin/GameOfLife.kt

```diff
@@ -0,0 +1,113 @@
+@file:Suppress("CanSealedSubClassBeObject")
+
+import GenerationCount.Finite
+import GenerationCount.Infinite
+import arrow.core.*
+import arrow.core.extensions.list.functorFilter.filterMap
+import arrow.core.extensions.list.semigroupal.times
+import arrow.fx.ForIO
+import arrow.fx.IO
+import arrow.fx.extensions.io.monad.monad
+import arrow.mtl.StateT
+
+sealed class Cell {
+    class Alive : Cell()
+    class Dead : Cell()
+}
+
+fun Cell.isDead() = !isAlive()
+fun Cell.isAlive() = this is Cell.Alive
+
+typealias Universe = List<List<Cell>>
+
+private val Universe.sizeX: Int get() = size
+
+private val Universe.sizeY: Int get() = get(0).size
+
+data class Position(val x: Int, val y: Int)
+
+fun Option<Position>.shiftBy(universe: Universe, delta: Tuple2<Int, Int>): Option<Position> =
+    flatMap { pos ->
+        val shifted = pos.copy(x = pos.x + delta.a, y = pos.y + delta.b)
+        if (shifted.x in 0 until universe.sizeX && shifted.y in 0 until universe.sizeY) {
+            shifted.some()
+        } else {
+            None
+        }
+    }
+
+fun Universe.cellPosition(cell: Cell): Option<Position> =
+    this.flatten().indexOf(cell).let { index ->
+        if (index == -1) {
+            None
+        } else {
+            val x = index / sizeX
+            val y = index % sizeX
+            Some(Position(x, y))
+        }
+    }
+
+fun Cell.neighbours(universe: Universe): List<Cell> {
```

Comment:
Maybe it could return the number of neighbours because it's not necessary to know the actual positions.

Reply:
I'm using this list in different ways from different places, e.g: getting all alive neighbours just by filtering this list. I could do the opposite to get the dead ones etc. Thought it'd be a nice to have utility function.

Reply:
Agree!!